### PR TITLE
Fix type annotations in pandas.core.indexes.datetimes

### LINKED ISSUE #26366: Fix type annotations for pandas.core.indexes.datetimes
Part of #25882

Current errors are:

    pandas\core\indexes\datetimes.py:7: error: Module 'pandas._libs' has no attribute 'join'
    pandas\core\indexes\datetimes.py:89: error: Definition of "_data" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:89: error: Definition of "_data" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "DatetimelikeDelegateMixin"
    pandas\core\indexes\datetimes.py:89: error: Definition of "map" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:89: error: Definition of "_format_with_header" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:89: error: Definition of "isin" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:1115: error: "Callable[[Any], Any]" has no attribute "fget"
    pandas\core\indexes\datetimes.py:1116: error: "Callable[[Any], Any]" has no attribute "fget"
    pandas\core\indexes\datetimes.py:1117: error: "Callable[[Any], Any]" has no attribute "fget"

### PR DESCRIPTION
**Pull request by vaibhavhrt**
- [x] closes #26366 
- [x] tests passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

`pandas/core/indexes/timedeltas.py` taken care of as well since the errors were the same.

### CODE REVIEW

**Comment by vaibhavhrt** [comment]
Errors that still remain are:

    pandas\core\indexes\datetimes.py:90: error: Definition of "_data" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:90: error: Definition of "_data" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "DatetimelikeDelegateMixin"
    pandas\core\indexes\datetimes.py:90: error: Definition of "map" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:90: error: Definition of "_format_with_header" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:90: error: Definition of "isin" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"

File: pandas/core/indexes/datetimes.py [hunk 1]
```diff
@@ -1112,9 +1113,11 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
...
     _is_unique = Index.is_unique
 
-    _timezone = cache_readonly(DatetimeArray._timezone.fget)
-    is_normalized = cache_readonly(DatetimeArray.is_normalized.fget)
-    _resolution = cache_readonly(DatetimeArray._resolution.fget)
+    _timezone = cache_readonly(DatetimeArray._timezone.fget)  # type: ignore
+    is_normalized = cache_readonly(
+        DatetimeArray.is_normalized.fget)  # type: ignore
+    _resolution = cache_readonly(
+        DatetimeArray._resolution.fget)  # type: ignore
```
Comment (vaibhavhrt) [inline pandas/core/indexes/datetimes.py]: There doesn't seems to be proper solution for `.fget` and `.fset` so ignoring it.

Comment (WillAyd) [reply pandas/core/indexes/datetimes.py]: Yea we had a similar conversation here:

https://github.com/pandas-dev/pandas/pull/26280#pullrequestreview-233722590

So I think OK for now though would love to find a better solution if you see anything else out there.

cc @jbrockmendel

**Review by WillAyd — COMMENTED** [review]
Looks like an isort failure in CI

File: pandas/core/indexes/datetimes.py (hunk 1 above)
Comment (vaibhavhrt) [reply pandas/core/indexes/datetimes.py]: I have subscribed to https://github.com/python/mypy/issues/6185 If some progress is made there I will get notified.

**Comment by vaibhavhrt** [comment]
@WillAyd some suggestion on what to do with these errors:

    pandas\core\indexes\datetimes.py:89: error: Definition of "_data" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:89: error: Definition of "_data" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "DatetimelikeDelegateMixin"
    pandas\core\indexes\datetimes.py:89: error: Definition of "map" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:89: error: Definition of "_format_with_header" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"
    pandas\core\indexes\datetimes.py:89: error: Definition of "isin" in base class "DatetimeIndexOpsMixin" is incompatible with definition in base class "Index"

List of attributes with conflicting types:

1. _data

    Class | Type | Initial Value
    ----- | ----- | -------------
    DatetimeIndexOpsMixin |  DatetimeLikeArrayMixin | None
    DatetimelikeDelegateMixin | None (using reveal_type) | None
    Index | None (using reveal_type) | None

    I don't think we should use `DatetimeLikeArrayMixin` in all three classes. Should I remove the type comment from class `DatetimeIndexOpsMixin` too?  
    Making it `Optional[DatetimeLikeArrayMixin]` in class `DatetimeIndexOpsMixin` will not fix the error.

2. map

    Class | Defintion
    ----- | ----------
    DatetimeIndexOpsMixin | def map(self, f):
    Index | def map(self, mapper, na_action=None):

3. _format_with_header

    Class | Defintion
    ----- | ----------
    DatetimeIndexOpsMixin | def _format_with_header(self, header, **kwargs):
    Index | def _format_with_header(self, header, na_rep='NaN', **kwargs):

4. isin

    Class | Defintion
    ----- | ----------
    DatetimeIndexOpsMixin | def isin(self, values):
    Index | def isin(self, values, level=None):

I don't see any easy way of fixing the methods which have different parameters. Maybe we should just ignore these errors too.

**Comment by WillAyd** [comment]
Haven’t been able to deeply review but you can try aligning the signatures for the methods - seems like the difference is only optional keywords anyway

**Comment by vaibhavhrt** [comment]
@WillAyd I can fix the last two but what about `def map(self, f)` and `def map(self, mapper, na_action=None)`. Is `f` and `mapper` the same thing?

**Comment by WillAyd** [comment]
I would think they are the same

**Comment by jreback** [comment]
this is ok, @vaibhavhrt you have some checks issue, can you fix up and ping on green

File: pandas/core/indexes/datetimelike.py [hunk 2]
```diff
@@ -430,8 +430,8 @@ def argmax(self, axis=None, skipna=True, *args, **kwargs):
     # --------------------------------------------------------------------
     # Rendering Methods
 
-    def _format_with_header(self, header, **kwargs):
-        return header + list(self._format_native_types(**kwargs))
+    def _format_with_header(self, header, na_rep='NaN', **kwargs):
```
Comment (vaibhavhrt) [inline pandas/core/indexes/datetimelike.py]: This line seems to be making tests fail. The default was `NaN` in class `Index` so I used `NaN` here too, but tests are expecting it to be `NaT`. I will change it to NaT and hope mypy doesn't complains about it.

**Review by vaibhavhrt — COMMENTED** [review]
Make tests pass!

**Comment by vaibhavhrt** [comment]
@WillAyd @jreback I have made all changes and tests are passing now. Please review.

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

**Comment by vaibhavhrt** [comment]
@jreback please review and merge.

File: mypy.ini [hunk 3]
```diff
@@ -8,11 +8,5 @@ ignore_errors=True
 [mypy-pandas.core.indexes.datetimelike]
 ignore_errors=True
 
-[mypy-pandas.core.indexes.datetimes]
-ignore_errors=True
-
 [mypy-pandas.core.indexes.period]
```
Comment (plammens) [inline mypy.ini]: I believe `pandas.core.indexes.period` can be removed too from the blacklist with these changes. I tried removing it and `mypy` reports no errors.

**Comment by plammens** [comment]
>    I don't think we should use `DatetimeLikeArrayMixin` in all three classes. Should I remove the type comment from class `DatetimeIndexOpsMixin` too?
>    Making it `Optional[DatetimeLikeArrayMixin]` in class `DatetimeIndexOpsMixin` will not fix the error.

I think `DatetimeLikeArrayMixin` for both `Datetime*` classes and `np.ndarray` for `Index` is also an option–it works, and it is the option that makes most sense to me. (Also because there is a comment that says https://github.com/pandas-dev/pandas/blob/65eac80dd4113ddbcee876029bfe03b4b3f11a3d/pandas/core/indexes/base.py#L522).

File: mypy.ini (hunk 3 above)
Comment (vaibhavhrt) [reply mypy.ini]: I will take a look, thanks for the heads up.

**Comment by vaibhavhrt** [comment]
@plammens there are a million different classes that inherits from the `Index` class and all those child classes have some different type for data.

File: pandas/core/indexes/datetimelike.py [hunk 4]
```diff
@@ -430,8 +430,8 @@ def argmax(self, axis=None, skipna=True, *args, **kwargs):
     # --------------------------------------------------------------------
     # Rendering Methods
 
-    def _format_with_header(self, header, **kwargs):
-        return header + list(self._format_native_types(**kwargs))
+    def _format_with_header(self, header, na_rep='NaT', **kwargs):
```
Comment (plammens) [inline pandas/core/indexes/datetimelike.py]: No big deal, but I think there's a typo (`'NaT'` instead of `'NaN'`). Or is it intentional?

Comment (jreback) [reply pandas/core/indexes/datetimelike.py]: this is correct, NaT is the datetime repr for missing values.

**Comment by jreback** [comment]
thanks @vaibhavhrt

File: pandas/core/indexes/datetimelike.py (hunk 4 above)
Comment (plammens) [reply pandas/core/indexes/datetimelike.py]: My bad!

File: mypy.ini (hunk 3 above)
Comment (vaibhavhrt) [reply mypy.ini]: Well there is just one error in `pandas.core.indexes.period`, but since this PR has been merged already, I created a new one #26518

**Pull request merged**